text/0000-consensus-api.md

```diff
@@ -0,0 +1,356 @@
+
+- All nodes are authorized to make commits at any given time.
+- Nodes need only be aware of their peers on the network (not all nodes).
+- Adding and removing nodes is trivial; the network supports "open-enrollment".
+- There are no "consensus-specific" messages; consensus is an emergent property
+  of the fork-resolution logic.
+
+Examples of _lottery_ consensus algorithms include Proof-of-Work (PoW)
+and Proof-of-Elapsed-Time (PoET).
+
+## Overview of the Consensus Engine
+
+The Consensus Engine lives in a separate process and interacts with the
+Validator using protobuf messages.
+
+The Consensus Engine is responsible for:
+
+- Determining messages to send to peers
+- Sending commands to progress the blockchain
+- Reacting to internal interrupts and a limited set of external messages
+
+The Consensus Engine is not responsible for:
+
+- Validating the integrity of blocks, batches, transactions
+- Validating block, batch, transaction, or message signatures
+- Gossiping blocks, batches, or transactions
+- Block storage, block creation, and (direct) management of the chain head
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Requirements
+
+The following is a list of necessary requirements for the Sawtooth
+architecture to be able to support both classes of consensus algorithms:
+
+**R1 - Consensus-Specific Messaging**
+
+At a minimum, an efficient communication channel for sending consensus messages
+between network nodes is needed. This channel must support both broadcast and
+peer-to-peer messaging.
+
+**R2 - Separate Consensus Process**
+
+Implementing consensus algorithms is expensive. It must be possible to reuse
+existing implementations to the extent possible. For this reason, it must be
+possible to run consensus in a separate process.
+
+**R3 - Consensus-Driven Block Publishing and Chain Updates**
+
+In order to be efficient, the consensus algorithm must be responsible for
+determining when to publish blocks and when to update the chain head.
+
+For example, in voting algorithms, only one node is authorized to
+publish blocks at any given time and the authorized node often does not change
+until it is deemed faulty. In a stable environment, this could be an indefinite
+amount of time. If consensus is not responsible for determining when to publish
+and update the chain head, all but one nodes will be polling consensus
+for long periods of time.
+
+**R4 - Read Access to Settings**
+
+Consensus engines need to read settings to support on-chain consensus engine
+configuration using the Settings Transaction Family. PoET depends on Settings
+to configure critical values such as target wait time and enclave measurements
+and these values must be consistent across the network.
+
+**R5 - Read/Write Access to Global State**
+
+PoET depends on tracking authorized block publishers in global state and
+therefore requires read access to a PoET specific namespace.
```

Comment:
Where does the "Write" (from "Read/Write Access") come in here.  I think you mean "read/write" access on line 128.

Reply:
This should just say Read